server/test/com/mirth/connect/server/userutil/DestinationSetTest.java

```diff
@@ -0,0 +1,408 @@
+
+        ImmutableConnectorMessage icm = new ImmutableConnectorMessage(connMsg, false, destinationIdMap);
+        return new DestinationSet(icm);
+    }
+
+    // @Test
+    // public void validateAssumptionsEmptySet() throws Exception {
+    //     HashSet<Integer> set = new HashSet<>();
+
+    //     assertFalse(set.addAll(Collections.emptyList()));
+
+    //     assertFalse(set.contains(null));
+
+    //     assertTrue(set.containsAll(Collections.emptyList()));
+    //     assertFalse(set.containsAll(Arrays.asList("Invalid")));
+    //     assertFalse(set.containsAll(Collections.singleton(null)));
+    // }
+
+    // @Test
+    // public void validateAssumptionsPopulatedSet() throws Exception {
+    //     HashSet<Integer> set = new HashSet<>(Arrays.asList(1,2,3,4,5));
+
+    //     assertFalse(set.addAll(Collections.emptyList()));
+
+    //     assertFalse(set.contains(null));
+
+    //     assertTrue(set.containsAll(Collections.emptyList()));
+    //     assertFalse(set.containsAll(Arrays.asList("Invalid")));
+    //     assertFalse(set.containsAll(Collections.singleton(null)));
+    // }
+
+    @Test
+    public void testIsEmptyWithoutSourceMapKey() throws Exception {
+        DestinationSet ds = getNewTestDestinationSetUsing(new Integer[0]);
+
+        assertTrue(ds.isEmpty());
+        assertEquals(0, ds.size());
+    }
+
+    @Test
+    public void testIsEmptyWithSourceMapKey() throws Exception {
+        ConnectorMessage cm = new ConnectorMessage();
+        Set<Integer> metaDataIds = new HashSet<>();
+        Map<String, Object> sm = new HashMap<>();
+        sm.put(Constants.DESTINATION_SET_KEY, metaDataIds);
+        cm.setSourceMap(sm);
+        ImmutableConnectorMessage icm = new ImmutableConnectorMessage(cm);
+        DestinationSet ds = new DestinationSet(icm);
+
+        assertTrue(ds.isEmpty());
+        assertEquals(0, ds.size());
+    }
+
+    @Test
+    public void testPopulatedSet() throws Exception {
+        DestinationSet ds = getNewTestDestinationSet();
+
+        assertFalse(ds.isEmpty());
+        assertEquals(3, ds.size());
+    }
+
+    @Test
+    public void testAdd() throws Exception {
+        DestinationSet ds = getNewTestDestinationSet();
+
+        assertTrue(ds.add(4));
+        assertEquals(4, ds.size());
+
+        //already in set
+        assertFalse(ds.add(1));
+        assertEquals(4, ds.size());
+
+        assertFalse(ds.add(null));
+        assertEquals(4, ds.size());
+    }
+
+    @Test
+    public void testAddAll() throws Exception {
+        DestinationSet ds = getNewTestDestinationSet();
+        
+        assertFalse(ds.addAll(Arrays.asList(1)));
+        assertEquals(3, ds.size());
+        assertFalse(ds.addAll(Arrays.asList(2, 3)));
+        assertEquals(3, ds.size());
+        assertTrue(ds.addAll(Arrays.asList(3, 4)));
+        assertEquals(4, ds.size());
+        assertTrue(ds.addAll(Arrays.asList(5, 6)));
+        assertEquals(6, ds.size());
+        //TODO is this what we want?
```

Comment:
I think so. The behavior of DestiationSet is such that if any operation fails, it returns `false`